src/actions/kloudio/kloudio.ts

```diff
@@ -0,0 +1,177 @@
+import * as https from "request-promise-native"
+
+import * as uuid from "uuid"
+import * as winston from "winston"
+import * as Hub from "../../hub"
+
+const signedUrl = "https://api.kloud.io/v1/tools/signed-url-put-object?key="
+const API_URL = "https://ecuoyevlzf.execute-api.us-west-2.amazonaws.com/prod/dest/send"
+const s3Bool = true
+let data = {}
+
+export class KloudioAction extends Hub.Action {
+
+  name = "kloudio"
+  label = "Kloudio"
+  iconName = "kloudio/kloudio.svg"
+  description = "Add records to a Google Spreadsheet."
+  params = []
+  supportedActionTypes = [Hub.ActionType.Query]
+  supportedFormats = [Hub.ActionFormat.JsonDetail]
+  supportedFormattings = [Hub.ActionFormatting.Unformatted]
+  supportedVisualizationFormattings = [Hub.ActionVisualizationFormatting.Noapply]
+
+  async execute(request: Hub.ActionRequest) {
+    if (!(request.attachment && request.attachment.dataJSON)) {
+        throw "No attached json."
+    }
+
+    if (!(request.formParams.apiKey)) {
+        throw "Missing API key"
+    }
+
+    if (!( request.formParams.url)) {
+        throw "Missing Google sheets URL"
+    }
+
+    const qr = request.attachment.dataJSON
+    if (!qr.fields || !qr.data) {
+      throw "Request payload is an invalid format."
+    }
+
+    winston.info(typeof request.attachment.dataJSON)
+
+    const { spreadsheetId, sheetId }  = await parseGsheet(request.formParams.url)
+
+    const labels = request.attachment.dataJSON.fields.dimensions.map((label: { label: any; }) => label.label)
+    const finalLabels = []
+    finalLabels.push(labels)
+    const names = request.attachment.dataJSON.fields.dimensions.map((labelId: { name: any; }) => labelId.name)
+    const dataRows = await parseData(request.attachment.dataJSON.data, names, finalLabels)
+
+    let response
+    const anonymousId = this.generateAnonymousId() + ".json"
+    const s3SignedUrl = await getS3Url(anonymousId, signedUrl, request.formParams.apiKey)
+
+    if (!s3SignedUrl.signedURL || s3SignedUrl.success === false) {
+      response = { success: false, message: "Invalid Kloudio API key" }
+      return new Hub.ActionResponse(response)
+    }
+
+    await uploadToS32(s3SignedUrl.signedURL, dataRows)
+    data = {destination: "looker", apiKey: request.formParams.apiKey, spreadsheetId , sheetId,
+       s3Upload: s3Bool, data: anonymousId, reportName: "Looker Report"}
+
+    try {
+        const newUri = API_URL.replace(/['"]+/g, "")
+        const lambdaResponse = await https.post({
+        url: newUri,
+        headers: {"Content-Type": "application/json"},
+        json: true,
+        body: data,
+         }).catch((_err) => {
+           const error = JSON.parse(_err)
+           winston.info("parsing error code" + error.emailCode)
```

Comment:
should this be an error level message? Also can this potentially log PII?

Reply:
Yes this has been made into an error level message. This does not log PII like user credentials or any other data. It logs generic error status like "GENERIC_ERROR" or "INVALID_PARAM"